src/plan/global.rs

```diff
@@ -253,30 +253,50 @@ pub trait Plan: 'static + Sync + Downcast {
      */
     fn collection_required(&self, space_full: bool, _space: &dyn Space<Self::VM>) -> bool;
 
-    fn get_pages_reserved(&self) -> usize {
-        self.get_pages_used() + self.get_collection_reserve()
+    /// Get the number of pages that are reserved, including used pages and pages that will
+    /// be used (e.g. for copying).
+    fn get_reserved_pages(&self) -> usize {
+        self.get_used_pages() + self.get_collection_reserved_pages()
     }
 
+    /// Get the total number of pages for the heap.
     fn get_total_pages(&self) -> usize {
         self.base().heap.get_total_pages()
     }
 
-    fn get_pages_avail(&self) -> usize {
-        self.get_total_pages() - self.get_pages_reserved()
-    }
-
-    fn get_collection_reserve(&self) -> usize {
+    /// Get the number of pages that are still available for use. The available pages
+    /// should always be positive or 0.
+    fn get_available_pages(&self) -> usize {
+        // It is possible that the reserved pages is larger than the total pages so we are doing
+        // a saturating substraction to make sure we return a non-negative number.
+        // For example,
+        // 1. our GC trigger checks if reserved pages is more than total pages.
+        // 2. when the heap is almost full of live objects and we are doing a copying GC, it is possible
+        //    the reserved pages is larger than total pages after the copying GC (the reserved pages after a GC
+        //    may be larger than the reserved pages before a GC, as we may end up using more memory for thread local
+        //    buffers for copy allocators).
```

Comment:
Can I suggest a minor edit to mention OOMs? Something like:
```
// 2. when the heap is almost full of live objects (such as in the case of an OOM) and we are doing a copying GC,
//    it is possible the reserved pages is larger than total pages after the copying GC (the reserved pages after
//    a GC may be larger than the reserved pages before a GC, as we may end up using more memory for thread local
//    buffers for copy allocators).
```

Reply:
Yeah. I will do that.